File: server/src/main/com/broll/gainea/server/core/battle/BattleContext.java
package com.broll.gainea.server.core.battle;

import com.broll.gainea.server.core.map.Location;
import com.broll.gainea.server.core.objects.Unit;
import com.broll.gainea.server.core.player.Player;
import com.broll.gainea.server.core.utils.PlayerUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

public class BattleContext {

    protected Location location;
    protected Location sourceLocation;
    protected List<Unit> attackers;
    protected List<Unit> defenders;
    protected Player attackingPlayer;
    protected List<Player> defendingPlayers;

    public BattleContext(List<Unit> attackers, List<Unit> defenders) {
        this.location = defenders.get(0).getLocation();
        this.sourceLocation = attackers.get(0).getLocation();
        this.attackers = attackers;
        this.defenders = defenders;
        this.attackingPlayer = PlayerUtils.getOwner(attackers);
        this.defendingPlayers = defenders.stream().map(Unit::getOwner).distinct().collect(Collectors.toList());
    }

    public Location getLocation() {
        return location;
    }

    public Location getSourceLocation() {
        return sourceLocation;
    }

    public Location getLocation(Player player) {
        if (isAttacker(player)) {
            return sourceLocation;
        } else if (isDefender(player)) {
            return location;
        }
        return null;
    }
    
    public List<Unit> getAttackers() {
        return attackers;
    }

    public List<Unit> getDefenders() {
        return defenders;
    }

    public List<Unit> getAliveAttackers() {
        return attackers.stream().filter(Unit::isAlive).collect(Collectors.toList());
    }

    public List<Unit> getAliveDefenders() {
        return defenders.stream().filter(Unit::isAlive).collect(Collectors.toList());
    }

    public List<Unit> getKilledAttackers() {
        return attackers.stream().filter(Unit::isDead).collect(Collectors.toList());
    }

    public List<Unit> getKilledDefenders() {
        return defenders.stream().filter(Unit::isDead).collect(Collectors.toList());
    }

    public boolean hasSurvivingAttackers() {
        return !getAliveAttackers().isEmpty();
    }

    public boolean hasSurvivingDefenders() {
        return !getAliveDefenders().isEmpty();
    }

    public List<Unit> getFightingArmy(Unit unit) {
        if (isAttacking(unit)) {
            return getAliveAttackers();
        } else if (isDefending(unit)) {
            return getAliveDefenders();
        }
        return new ArrayList<>();
    }

    public List<Unit> getUnits(Player player) {
        if (isAttacker(player)) {
            return attackers.stream().filter(it -> it.getOwner() == player).collect(Collectors.toList());
        } else if (isDefender(player)) {
            return defenders.stream().filter(it -> it.getOwner() == player).collect(Collectors.toList());
        }
        return new ArrayList<>();
    }

    public List<Unit> getOpposingUnits(Player player) {
        if (isAttacker(player)) {
            return getDefenders();
        } else if (isDefender(player)) {
            return getAttackers();
        }
        return new ArrayList<>();
    }

    public List<Unit> getOpposingFightingArmy(Unit unit) {
        if (isAttacking(unit)) {
            return getAliveDefenders();
        } else if (isDefending(unit)) {
            return getAliveAttackers();
        }
        return new ArrayList<>();
    }

    public boolean isParticipating(Player player) {
        return isAttacker(player) || isDefender(player);
    }

    public boolean isAttacker(Player player) {
        return this.attackingPlayer == player;
    }

    public boolean isAttacking(Unit unit) {
        return this.attackers.contains(unit);
    }

    public boolean isDefending(Unit unit) {
        return this.defenders.contains(unit);
    }

    public boolean isDefender(Player player) {
        return this.defendingPlayers.contains(player);
    }

    public boolean isNeutralAttacker() {
        return this.attackingPlayer == null;
    }

    public boolean isNeutralDefender() {
        return isNeutralOwner(this.defendingPlayers);
    }

    public boolean isNeutralParticipant() {
        return isNeutralAttacker() || isNeutralDefender();
    }

    public Player getAttackingPlayer() {
        return attackingPlayer;
    }

    protected boolean isNeutralOwner(List<Player> players) {
        return players.size() == 1 && players.get(0) == null;
    }

    protected List<Player> getNonNeutralOwners(List<Player> owners) {
        return owners.stream().filter(Objects::nonNull).collect(Collectors.toList());
    }

    public List<Player> getDefendingPlayers() {
        return getNonNeutralOwners(this.defendingPlayers);
    }
}

File: server/src/main/com/broll/gainea/server/core/battle/BattleContext.kt
package com.broll.gainea.server.core.battle

import com.broll.gainea.server.core.map.Location
import com.broll.gainea.server.core.objects.IUnit
import com.broll.gainea.server.core.objects.Unit
import com.broll.gainea.server.core.player.Player
import com.broll.gainea.server.core.player.isNeutral


abstract class AbstractBattleContext<U : IUnit>(
    attackers: List<U>,
    defenders: List<U>,
    val rounds: MutableList<BattleRound> = mutableListOf()
) {

    var attackers: List<U> = attackers
        protected set
    var defenders: List<U> = defenders
        protected set

    val location: Location
        get() = defenders[0].location
    val sourceLocation: Location
        get() = attackers[0].location

    val attackingPlayers: List<Player>
        get() = attackers.map { it.owner }.distinct()
    val defendingPlayers: List<Player>
        get() = defenders.map { it.owner }.distinct()


    fun getLocation(player: Player): Location? {
        if (isAttacker(player)) {
            return sourceLocation
        } else if (isDefender(player)) {
            return location
        }
        return null
    }

    val aliveAttackers: List<U>
        get() = attackers.filter { it.alive }
    val aliveDefenders: List<U>
        get() = defenders.filter { it.alive }
    val killedAttackers: List<U>
        get() = attackers.filter { it.dead }
    val killedDefenders: List<U>
        get() = defenders.filter { it.dead }

    fun hasSurvivingAttackers() = aliveAttackers.isNotEmpty()

    fun hasSurvivingDefenders() = aliveDefenders.isNotEmpty()

    fun getFightingArmy(unit: IUnit): List<U> {
        if (isAttacking(unit)) {
            return aliveAttackers
        } else if (isDefending(unit)) {
            return aliveDefenders
        }
        return listOf()
    }

    fun getUnits(player: Player): List<U> {
        if (isAttacker(player)) {
            return attackers.filter { it.owner == player }
        } else if (isDefender(player)) {
            return defenders.filter { it.owner == player }
        }
        return listOf()
    }

    fun getOpposingUnits(player: Player): List<U> {
        if (isAttacker(player)) {
            return defenders
        } else if (isDefender(player)) {
            return attackers
        }
        return listOf()
    }

    fun getOpposingFightingArmy(unit: IUnit): List<U> {
        if (isAttacking(unit)) {
            return aliveDefenders
        } else if (isDefending(unit)) {
            return aliveAttackers
        }
        return listOf()
    }

    fun isParticipating(player: Player) = isAttacker(player) || isDefender(player)

    fun isParticipating(unit: IUnit) = attackers.contains(unit) || defenders.contains(unit)

    fun isAttacker(player: Player) = attackingPlayers.contains(player)


    fun isAttacking(unit: IUnit) = attackers.contains(unit)


    fun isDefending(unit: IUnit) = defenders.contains(unit)


    fun isDefender(player: Player) = defendingPlayers.contains(player)


    val isNeutralAttacker: Boolean
        get() = isNeutralOwner(attackingPlayers)
    val isNeutralDefender: Boolean
        get() = isNeutralOwner(defendingPlayers)
    val isNeutralParticipant: Boolean
        get() = isNeutralAttacker || isNeutralDefender

    protected fun isNeutralOwner(players: List<Player>): Boolean {
        return players.size == 1 && players[0].isNeutral()
    }

    protected fun getNonNeutralOwners(owners: List<Player>) = owners.filterNot { it.isNeutral() }

    fun getNonNeutralAttackers() =
        getNonNeutralOwners(attackingPlayers)

    fun getNonNeutralDefenders() =
        getNonNeutralOwners(defendingPlayers)

    fun getControllingAttacker() = getNonNeutralAttackers().firstOrNull()

}

data class BattleRound(val damageTaken: List<FightResult.AttackDamage>)

class BattleContext(attackers: List<Unit>, defenders: List<Unit>) :
    AbstractBattleContext<Unit>(attackers, defenders) {

    fun addAttacker(unit: Unit) {
        this.attackers = this.attackers.toMutableList().also { it.add(unit) }
    }

    fun addDefender(unit: Unit) {
        this.defenders = this.defenders.toMutableList().also { it.add(unit) }
    }

}